src/main/java/com/aws/greengrass/authorization/WildcardVariableTrie.java

```diff
@@ -0,0 +1,245 @@
+                current = current.children.get(String.valueOf(GLOB_WILDCARD));
+                current.isWildcard = true;
+                // If the string ends with *, then the wildcard is a terminal
+                if (i == subject.length() - 1) {
+                    current.isTerminal = isTerminal;
+                    return current;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            if (subject.charAt(i) == MQTT_MULTILEVEL_WILDCARD) {
+                WildcardVariableTrie terminalLevel = current.add(sb.toString(), false);
+                current = terminalLevel.children.get(String.valueOf(MQTT_MULTILEVEL_WILDCARD));
+                if (i == subject.length() - 1) {
+                    current.isTerminal = true;
+                    // check if # wildcard usage is valid
+                    if (i > 0 && subject.charAt(i - 1) == MQTT_LEVEL_SEPARATOR) {
+                        current.isMQTTWildcard = true;
+                        current.matchAll = true;
+                        terminalLevel.isTerminalLevel = true;
+                    }
+                    return current;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            if (subject.charAt(i) == MQTT_SINGLELEVEL_WILDCARD) {
+                current = current.add(sb.toString(), false);
+                current = current.children.get(String.valueOf(MQTT_SINGLELEVEL_WILDCARD));
+                if (i == subject.length() - 1) {
+                    current.isTerminal = true;
+                    // check if '+' wildcard usage is valid
+                    // if it's used at the last level
+                    if (i > 0 && subject.charAt(i - 1) == MQTT_LEVEL_SEPARATOR) {
+                        current.isMQTTWildcard = true;
+                    }
+                    return current;
+                }
+                // check if '+' wildcard usage is valid
+                // if it's used in middle levels
+                if (i > 0 && subject.charAt(i - 1) == MQTT_LEVEL_SEPARATOR
+                        && subject.charAt(i + 1) == MQTT_LEVEL_SEPARATOR) {
+                    current.isMQTTWildcard = true;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            sb.append(subject.charAt(i));
+        }
+        // Handle non-wildcard value
+        current = current.children.get(sb.toString());
+        current.isTerminal |= isTerminal;
+        return current;
+    }
+
+    /**
+     * Match given string to the corresponding allowed resources trie. MQTT wildcards are processed only if
+     * allowMQTT is true.
+     *
+     * @param str string to match
+     * @param allowMQTT If MQTT wildcards are allowed to match
+     */
+    @SuppressWarnings({"PMD.UselessParentheses", "PMD.CollapsibleIfStatements"})
+    public boolean matches(String str, boolean allowMQTT) {
```

Comment:
should you split this whole method by allowMQTT? Would that make it easier to reason about? Would it improve performance? Have `matchesMqtt` and `matchesStandard`

Reply:
Yes, most of the code would be duplicated but I think makes it lot more readable